### PR TITLE
Radio jammer component

### PR DESCRIPTION
**Pull request by PowerfulBacon**
## About The Pull Request

Converts radio jamming to a component.

## Why It's Good For The Game

This provides no player facing changes.
This allows coders to create devices which act as radio jammers without requiring them to be a subtype of /obj/item/jammer. Requiring subtypes for functionalitites is bad, since if you have 2 things that work this way, you cannot create an item that combines the functionality of both without duplicating code, which is bad for maintainability.

## Testing Photographs and Procedure

image
Fixed the message being inverted

## Changelog
:cl:
refactor: Radio jamming is now component based. (No player facing changes.)
/:cl:

### CODE REVIEW

File: code/game/objects/items/devices/traitordevices.dm [hunk 1]
```diff
@@ -278,22 +278,19 @@ effective or pretty fucking useless.
...
-	else
-		GLOB.active_jammers -= src
-	update_icon()
+	SEND_SIGNAL(src, COMSIG_TOGGLE_JAMMER, user, FALSE)
 
 /atom/proc/is_jammed()
 	var/turf/position = get_turf(src)
-	for(var/obj/item/jammer/jammer in GLOB.active_jammers)
-		var/turf/jammer_turf = get_turf(jammer)
+	for(var/datum/component/radio_jamming/jammer in GLOB.active_jammers)
```
Comment (Zonespace27) [inline code/game/objects/items/devices/traitordevices.dm]:
```suggestion
	for(var/datum/component/radio_jamming/jammer as anything in GLOB.active_jammers)
```

File: code/datums/components/radio_jamming.dm [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+/datum/component/radio_jamming/proc/toggle(datum/source, mob/user, silent = FALSE)
+	SIGNAL_HANDLER
+	//Toggle the jammer
+	if (active)
+		disable(source, user, silent)
+	else
+		enable(source, user, silent)
+	if (!silent && user)
+		to_chat(user,"<span class='notice'>You [!active ? "deactivate" : "activate"] [parent].</span>")
```
Comment (Zonespace27) [inline code/datums/components/radio_jamming.dm]:
```suggestion
		to_chat(user, "<span class='notice'>You [!active ? "deactivate" : "activate"] [parent].</span>")
```

Comment (PowerfulBacon) [reply code/datums/components/radio_jamming.dm]: Good spot

File: code/game/objects/items/devices/traitordevices.dm (hunk 1 above)
Comment (PowerfulBacon) [reply code/game/objects/items/devices/traitordevices.dm]: There should be no way for this to hard del and stay null, so this should be fine

File: code/datums/components/radio_jamming.dm [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+	if (!active)
+		return
+	active = FALSE
+	GLOB.active_jammers -= src
+
+/datum/component/radio_jamming/proc/toggle(datum/source, mob/user, silent = FALSE)
+	SIGNAL_HANDLER
+	//Toggle the jammer
+	if (active)
+		disable(source, user, silent)
```
Comment (itsmeow) [inline code/datums/components/radio_jamming.dm]: these methods don't have these args

**Review by itsmeow — CHANGES_REQUESTED** [review]

File: code/datums/components/radio_jamming.dm [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+	else
+		enable(source, user, silent)
```
Comment (PowerfulBacon) [inline code/datums/components/radio_jamming.dm]:
```suggestion
		disable()
	else
		enable()
```
They used to

**Review by itsmeow — APPROVED** [review]
Looks good although how would someone pass a range to the component's Initialize?

**Comment by PowerfulBacon** [comment]
The same was you pass parameters to any component
AddComponent(/datum/component/whatever, 50)

**Review by Penwin0 — APPROVED** [review]

**Pull request merged**